### PR TITLE
Yapf RFC

### PR DESCRIPTION
**Pull request by dannon**
Yapf is a tool similar to Prettier in that it reads code in and completely rewrites it with defined rules, instead of simply trying to fix particular errors like autopep8 and the like.

I think most people really like the behavior in Javascript, so it was mentioned in IRC that this might be a great tool to start using with our python codebase as well.   Here's what that'd look like.

I ran this initial test with pep8 configuration and an arbitrarily chosen max_line_length of 160.  There are lots of 'knobs' though, that we can use, specified here: https://pypi.python.org/pypi/yapf#id8

Thoughts?

### CODE REVIEW

**Comment by jmchilton** [comment]
I thought I'd like this idea but there is definitely some stuff in there I hate. I feel like a big reason to use these tools is reduce the likelihood of conflicts during PRs and ease merging during conflicts. But the way it is flattening everything in the following slightly different examples makes things worse in terms of code going stale and make things harder for git to auto-merge.

```diff
-from galaxy.tools.cache import (
- ToolCache,
- ToolShedRepositoryCache
-)
+from galaxy.tools.cache import (ToolCache, ToolShedRepositoryCache)
```

```diff
-__all__ = (
- 'check_binary',
- 'check_bz2',
- 'check_gzip',
- 'check_html',
- 'check_image',
- 'check_zip',
- 'is_gzip',
- 'is_bz2',
-)
+__all__ = ('check_binary', 'check_bz2', 'check_gzip', 'check_html', 'check_image', 'check_zip', 'is_gzip', 'is_bz2', )
```

```diff
right_association = DatasetCollectionElement(
element=HistoryDatasetAssociation(),
- element_identifier=REVERSE_IDENTIFIER,
- )
+ element_identifier=REVERSE_IDENTIFIER, )
```

Update: I reviewed the available knobs and I don't see a solution for this.

**Comment by martenson** [comment]
Looking at first 10 files or so of this PR there is fairly large number of changes I perceive as readability regressions. I can comment inline for the particular cases if asked.

**Comment by dannon** [comment]
Yeah, I really like the idea of the completely automated code formatter, and I think there are a ton of improvements here, but I think dialing in the right set of rules that work for us is key.  I'm looking through the knobs in question to see what we can tweak for particular cases, if folks want to point them out like @jmchilton has.

File: lib/galaxy/actions/admin.py [hunk 1]
```diff
@@ -60,14 +66,16 @@ def _create_quota(self, params, decode_id=None):
                 for group in in_groups:
                     gqa = self.app.model.GroupQuotaAssociation(group, quota)
                     self.sa_session.add(gqa)
-                message = "Quota '%s' has been created with %d associated users and %d associated groups." % (quota.name, len(in_users), len(in_groups))
+                message = "Quota '%s' has been created with %d associated users and %d associated groups." % (quota.name, len(in_users),
```
Comment (martenson) [inline lib/galaxy/actions/admin.py]: cutting off tails like this feels less readable

File: lib/galaxy/app.py [hunk 2]
```diff
@@ -22,28 +22,22 @@
 from galaxy.sample_tracking import external_service_types
 from galaxy.openid.providers import OpenIDProviders
 from galaxy.tools.data_manager.manager import DataManagers
-from galaxy.tools.cache import (
-    ToolCache,
-    ToolShedRepositoryCache
-)
+from galaxy.tools.cache import (ToolCache, ToolShedRepositoryCache)
```
Comment (martenson) [inline lib/galaxy/app.py:25]: this style of imports breaks the rule of atomicity, which I like

File: lib/galaxy/app.py [hunk 3]
```diff
@@ -132,18 +126,15 @@ def __init__(self, **kwargs):
         load_lib_tools(self.toolbox)
         # visualizations registry: associates resources with visualizations, controls how to render
         self.visualizations_registry = VisualizationsRegistry(
-            self,
-            directories_setting=self.config.visualization_plugins_directory,
-            template_cache_dir=self.config.template_cache)
+            self, directories_setting=self.config.visualization_plugins_directory, template_cache_dir=self.config.template_cache)
```
Comment (martenson) [inline lib/galaxy/app.py]: this line is harder to read than our version of it

**Comment by dannon** [comment]
@jmchilton How often do you think those particular examples would cause conflicts?  I think it's likely still a net positive in that we'll have fewer conflicts (because no style clashes again, ever!), and those particular examples would be simple resolutions.

(dedent trailing option actually does resolve one of them, too, but I'm not sure I love it)

File: lib/galaxy/auth/providers/ldap_ad.py [hunk 4]
```diff
@@ -125,18 +126,19 @@ def authenticate(self, email, username, password, options):
...
-                                    _get_subs(options, 'search-password', params))
+                    l.simple_bind_s(_get_subs(options, 'search-user', params), _get_subs(options, 'search-password', params))
                 else:
                     l.simple_bind_s()
 
                 # setup search
-                attributes = [_.strip().format(**params)
-                              for _ in options['search-fields'].split(',')]
-                suser = l.search_ext_s(_get_subs(options, 'search-base', params),
+                attributes = [_.strip().format(**params) for _ in options['search-fields'].split(',')]
```
Comment (martenson) [inline lib/galaxy/auth/providers/ldap_ad.py]: it makes the line longer here, but does not improve readability I think

File: lib/galaxy/containers/docker.py [hunk 5]
```diff
@@ -29,14 +28,38 @@ class DockerInterface(ContainerInterface):
...
         # `run` options
-        'environment': {'flag': '--env', 'type': 'list_of_kvpairs'},
-        'volumes': {'flag': '--volume', 'type': 'docker_volumes'},
-        'name': {'flag': '--name', 'type': 'string'},
-        'detach': {'flag': '--detach', 'type': 'boolean'},
-        'publish_all_ports': {'flag': '--publish-all', 'type': 'boolean'},
-        'publish_port_random': {'flag': '--publish', 'type': 'string'},
-        'cpus': {'flag': '--cpus', 'type': 'string'},
-        'memory': {'flag': '--memory', 'type': 'string'},
+        'environment': {
```
Comment (martenson) [inline lib/galaxy/containers/docker.py:31]: I am ambivalent here, I guess ok?

File: lib/galaxy/datatypes/blast.py [hunk 6]
```diff
@@ -253,16 +251,30 @@ def __init__(self, **kwd):
...
-        self.add_composite_file('blastdb.nni', is_binary=True, optional=True)  # index of GI values ( -parse_seqids option of makeblastdb and gi present in the description lines)
-        self.add_composite_file('blastdb.nog', is_binary=True, optional=True)  # OID->GI lookup file ( -hash_index or -parse_seqids option of makeblastdb)
-        self.add_composite_file('blastdb.nsd', is_binary=True, optional=True)  # sorted sequence accession values ( -hash_index or -parse_seqids option of makeblastdb)
-        self.add_composite_file('blastdb.nsi', is_binary=True, optional=True)  # index of sequence accession values ( -hash_index or -parse_seqids option of makeblastdb)
-#        self.add_composite_file('blastdb.00.idx', is_binary=True, optional=True)  # first volume of the MegaBLAST index generated by makembindex
-# The previous line should be repeated for each index volume, with filename extensions like '.01.idx', '.02.idx', etc.
-        self.add_composite_file('blastdb.shd', is_binary=True, optional=True)  # MegaBLAST index superheader (-old_style_index false option of makembindex)
+        self.add_composite_file(
```
Comment (martenson) [inline lib/galaxy/datatypes/blast.py:270]: our version makes much more sense

File: lib/galaxy/managers/libraries.py [hunk 7]
```diff
@@ -206,14 +209,19 @@ def get_current_roles(self, trans, library):
...
         """
-        access_library_role_list = [(access_role.name, trans.security.encode_id(access_role.id)) for access_role in self.get_access_roles(trans, library)]
-        modify_library_role_list = [(modify_role.name, trans.security.encode_id(modify_role.id)) for modify_role in self.get_modify_roles(trans, library)]
-        manage_library_role_list = [(manage_role.name, trans.security.encode_id(manage_role.id)) for manage_role in self.get_manage_roles(trans, library)]
-        add_library_item_role_list = [(add_role.name, trans.security.encode_id(add_role.id)) for add_role in self.get_add_roles(trans, library)]
-        return dict(access_library_role_list=access_library_role_list,
-                    modify_library_role_list=modify_library_role_list,
-                    manage_library_role_list=manage_library_role_list,
-                    add_library_item_role_list=add_library_item_role_list)
+        access_library_role_list = [(access_role.name, trans.security.encode_id(access_role.id))
```
Comment (martenson) [inline lib/galaxy/managers/libraries.py]: the readability degraded here

**Review by martenson — COMMENTED** [review]
some random finds that I think are not useful or suboptimal

**Comment by jxtx** [comment]
A lot of the regressions seem to be making longer lines. Maybe really be aggressive on maximum line length? 

imports are a problem I agree. Maybe a showstopper.

**Comment by jmchilton** [comment]
@dannon I've been rebasing the same CWL branch against Galaxy changes for almost 2 years and by far the most common conflicts are things like this. Take the ``__all__`` example for instance - if everything is grouped together like that and I have added an export both to dev and my branch that line will always cause a conflict. If they are each on their own line they are much less likely to conflict and on the off change there is a conflict the resolution to the conflict is much easier since I know to keep both lines. Same applies to complex function calls, complex function definitions, and complex imports.

**Comment by dannon** [comment]
@jmchilton Makes sense with that type of rebase (long running cwl fork), I understand the problem for you now, thanks.  Most of my past conflict and pr review issues (weeding through large PRs to figure out what's a style change, what's a substantive one, etc) would be resolved with this, but admittedly I don't have a long running continually rebased fork like you do.

Let me see what line length does here.

**Comment by dannon** [comment]
Pushed an update to enforce 80 with the basic pep8 rules, after trying out both 'google' and 'facebook' style.  I thought I'd hate the 80 characters, but I actually don't mind it.  It is what everyone else on the planet uses, after all.

**Comment by martenson** [comment]
>  It is what everyone else on the planet uses, after all.

I do not believe we should use that as an argument, since "everyone on the planet uses 2-indented javascript" ( except of jQuery that uses TABs, lol)

**Comment by dannon** [comment]
@martenson I'm saying that it doesn't bother me as much as I expected, possibly because all the other code I read is this way, too.

**Comment by jmchilton** [comment]
I had previously liked all the changes in ``admin/actions/__init__.py`` with the longer line lengths and now I dislike them - I feel like shorter lines actually made this worse? I don't think most Python projects actually follow the 80 character guideline to be honest - I used to try for Pulsar and Planemo but I had to abandon it because it wasn't practical - it consistently hurt readability.

> Most of my past conflict and pr review issues (weeding through large PRs to figure out what's a style change, what's a substantive one, etc) would be resolved with this

Agreed this is a problem and we should continue to work toward improving it. Part of it is cultural - asking separation of style and refactoring PRs from PRs that add features. Part of it could be fixed with tools that don't rewrite ASTs but simply fix up problems - like autopep8?

Certainly rewriting the AST it is great for Go and it seems somewhat liked by the JavaScript community - but that doesn't mean it is right for Python right? I read a blog post earlier about this and they made the argument that since you already have to care so much about indentation. It seemed like a very big logical leap to me - but I'm starting to think there is something about Python that doesn't lend itself to doing this.

Update: Here is the blog post - https://utcc.utoronto.ca/~cks/space/blog/python/PythonNoFullGofmt - I don't necessarily think I agree with the logic but it is another perspective.

**Comment by dannon** [comment]
Thanks for the blog post link.  I agree with the post that spacing and formatting in python feels more pesronal (which is probably a bad thing, at least in a large project?).

To me, I like some of the changes in this PR, I hate a few, but the important point and what outweighs it all for me is that it's exactly the same formatting everywhere, regardless of who writes the code or when it was written.  I'm still leaning towards this being a good move, and that we just need to figure out the right way to apply this sort of pass, but I'll put more thought into it.

**Comment by jxtx** [comment]
I think this is looking like a non-starter at this time. We should table this. If we do want to do it eventually I think it needs to come at a time when we don't have so many big PRs/branches in progress. It was a worthwhile experiment.

**Comment by dannon** [comment]
@jxtx Sure, revisiting it down the road sounds fine to me.  Nothing urgent here.

**Comment by jmchilton** [comment]
> It was a worthwhile experiment.

Agreed whole heartedly - I really thought I'd like this. This was a great experiment and many thanks for checking this out @dannon.

Another thing to think about when we revisit this in the future is that Pulsar/Planemo/bioblend/galaxy-lib all have fewer contributors, fewer outstanding forks, and fewer lines of code - perhaps exploring the ideas there and then brining them to Galaxy proper when people are more comfortable is a good idea? I know that we have followed that pattern with various linting and testing things in the past and it has proven a good pattern.

**Pull request closed without merging**